packages/core/src/indexing-store/historical.ts

```diff
@@ -1,363 +1,893 @@
 import { HeadlessKysely } from "@/database/kysely.js";
 import type { NamespaceInfo } from "@/database/service.js";
 import type { Schema, Table } from "@/schema/common.js";
+import {
+  getTables,
+  isEnumColumn,
+  isJSONColumn,
+  isOptionalColumn,
+  isReferenceColumn,
+  isScalarColumn,
+} from "@/schema/utils.js";
 import type {
   DatabaseColumn,
   DatabaseRecord,
   UserId,
   UserRecord,
 } from "@/types/schema.js";
+import { createQueue } from "@ponder/common";
 import { sql } from "kysely";
-import type { WhereInput, WriteStore } from "./store.js";
-import { decodeRow, encodeRow, encodeValue } from "./utils/encoding.js";
+import { type Hex, padHex } from "viem";
+import { getReadonlyStore } from "./readonly.js";
+import type { HistoricalStore, OrderByInput, WhereInput } from "./store.js";
+import {
+  decodeRecord,
+  encodeField,
+  encodeRecord,
+  validateRecord,
+} from "./utils/encoding.js";
 import { parseStoreError } from "./utils/errors.js";
 import { buildWhereConditions } from "./utils/filter.js";
 
-const MAX_BATCH_SIZE = 1_000 as const;
+const MAX_BATCH_SIZE = 1_000;
+const CACHE_FLUSH = 0.35;
+
+type Entry<record = UserRecord> = {
+  opIndex: number;
+  bytes: number;
+  record: record;
+};
+
+type StoreCache = {
+  [tableName: string]: {
+    insert: { [id: string | number]: Entry };
+    update: { [id: string | number]: Entry };
+    find: { [id: string | number]: Entry<UserRecord | undefined> };
+  };
+};
 
 export const getHistoricalStore = ({
-  kind,
+  encoding,
   schema,
   namespaceInfo,
   db,
+  common,
+  isCacheFull: _isCacheFull,
 }: {
-  kind: "sqlite" | "postgres";
+  encoding: "sqlite" | "postgres";
   schema: Schema;
   namespaceInfo: NamespaceInfo;
   db: HeadlessKysely<any>;
-}): WriteStore<"historical"> => ({
-  create: async ({
-    tableName,
-    id,
-    data = {},
-  }: {
-    tableName: string;
-    id: UserId;
-    data?: Omit<UserRecord, "id">;
-  }) => {
-    const table = (schema[tableName] as { table: Table }).table;
-
-    return db.wrap({ method: `${tableName}.create` }, async () => {
-      const createRow = encodeRow({ id, ...data }, table, kind);
-
-      const row = await db
-        .withSchema(namespaceInfo.userNamespace)
-        .insertInto(tableName)
-        .values(createRow)
-        .returningAll()
-        .executeTakeFirstOrThrow()
-        .catch((err) => {
-          throw parseStoreError(err, { id, ...data });
-        });
+  common: Common;
+  isCacheFull: boolean;
+}): HistoricalStore => {
+  const maxSizeBytes = common.options.indexingCacheBytes;
+  const storeCache: StoreCache = {};
+  const tables = getTables(schema);
+  const readonlyStore = getReadonlyStore({
+    encoding,
+    schema,
+    namespaceInfo,
+    db,
+  });
+
+  common.logger.debug({
+    service: "indexing",
+    msg: `Using a ${Math.round(maxSizeBytes / (1024 * 1024))} mB cache.`,
+  });
+
+  /** True if the cache contains the complete state of the store. */
+  let isCacheFull = _isCacheFull;
```

Comment:
`isCacheComplete` may be better - "full" makes it seem like it's full and needs to be flushed.